src/utils/allocators.zig

```diff
@@ -514,6 +520,184 @@ pub fn RecycleFBA(config: struct {
+                .rawAlloc(len_with_offset, ptr_align_with_offset, ret_addr)) |ptr|
+            {
+                writeBatchPtr(batch, ptr[0..len_with_offset]);
+                return ptr;
+            } else {
+                _ = batch.num_allocs.fetchSub(1, .monotonic);
+            }
+        }
+        return null;
+    }
+
+    fn tryAllocNewBatch(
+        self: *Self,
+        len_with_offset: usize,
+        ptr_align_with_offset: u8,
+        ret_addr: usize,
+    ) ?[*]u8 {
+        // only one thread should do this at a time.
+        // other threads should wait until the first one finishes.
+        if (self.new_batch_waiters.fetchAdd(1, .monotonic) > 0) {
+            _ = self.new_batch_waiters.fetchSub(1, .monotonic);
+            std.atomic.spinLoopHint();
+            return null;
+        }
+        defer _ = self.new_batch_waiters.fetchSub(1, .monotonic);
+        self.new_batch_lock.lock();
+        defer self.new_batch_lock.unlock();
+
+        // create batch
+        const padding = @max(@as(usize, 1) << @intCast(ptr_align_with_offset), @sizeOf(Batch));
+        const batch_size = @max(self.batch_size, len_with_offset + padding);
+        const batch_bytes: [*]align(@alignOf(Batch)) u8 = @alignCast(
+            self.backing_allocator.rawAlloc(batch_size, log2(@alignOf(Batch)), ret_addr) orelse
+                return null,
+        );
+        const new_batch: *Batch = @ptrCast(batch_bytes);
+        const buffer: []u8 = batch_bytes[@sizeOf(Batch)..batch_size];
+        new_batch.* = Batch{
+            .ptr = batch_bytes,
+            .len = batch_size,
+            .fba = FixedBufferAllocator.init(buffer),
+            .num_allocs = Atomic(usize).init(1),
+        };
+
+        // use new batch for allocation
+        const ptr = new_batch.fba.threadSafeAllocator()
+            .rawAlloc(len_with_offset, ptr_align_with_offset, ret_addr) orelse unreachable;
+        writeBatchPtr(new_batch, ptr[0..len_with_offset]);
+
+        if (len_with_offset >= self.batch_size) {
+            // this batch won't be useful for any other allocations
+            _ = new_batch.num_allocs.fetchOr(Batch.DONE, .monotonic);
+        } else {
+            // make new batch available to future allocations, and mark old batch as done
+            if (self.last_batch.swap(new_batch, .monotonic)) |old_batch| {
+                if (0 == old_batch.num_allocs.fetchOr(Batch.DONE, .monotonic)) {
+                    // free old batch since all allocations are freed
+                    old_batch.deinit(self.backing_allocator);
+                }
+            }
+        }
+
+        return ptr;
+    }
+
+    fn writeBatchPtr(batch: *Batch, slice_with_batch: []u8) void {
+        const batch_int = @intFromPtr(batch);
+        const batch_ptr_bytes: [@sizeOf(usize)]u8 = @bitCast(batch_int);
+        @memcpy(
+            slice_with_batch[slice_with_batch.len - @sizeOf(usize) .. slice_with_batch.len],
+            &batch_ptr_bytes,
+        );
```

Comment:
```suggestion
        const batch_ptr_slice = slice_with_batch[slice_with_batch.len - @sizeOf(usize)..][0..@sizeOf(*Batch)];
        std.mem.bytesAsValue(*Batch, batch_ptr_slice) = batch;
```

Reply:
It's not clear to me why `bytesAsValue` exists in std when `@ptrCast` already exists. `bytesAsValue` is less widely used, which makes it esoteric. In my opinion, this makes the code harder to understand compared to using `@ptrCast` directly. What do you think?

There is also a change in semantics with your suggestion. Previously, this was just copying the bytes of the pointer. In that case, it doesn't matter how the bytes are aligned. With your suggestion, the bytes in the allocation are directly interpreted as the pointer. This means the bytes need to be properly aligned.

To align the bytes properly, we need some new logic to adjust the size and alignment of allocation. Also, the `writeBatchPtr` and `getBatch` functions need to change in tandem. I also ended up refactoring this a bit to consolidate/encapsulate this logic and clarify the meaning of the inputs. Hopefully the code is easy to understand.

03c83a4cf3e4baf994783a93a6e3eca6bea0d36f

Reply:
`bytesAsValue` directly copies the pointer attributes of the source byte ptr - so for example, in this case, the pointer it returns will be of type `*align(1) *Batch`. It results in the same fundamental operation as was being done here, ie it copies the bytes representing the pointer to the slice, which ends up being an unaligned copy; the benefit here is the type safety (ie avoiding `@intFromPtr`, and automatically inferring the appropriate pointer alias properties).

The change you've applied does improve things, but it does incur potentially more memory usage than my suggestion, since it increases the alignment requirement for storing metadata.

I'm fine to accept the changes you've made, but perhaps consider them in light of my explanation.

Reply:
> it does incur potentially more memory usage than my suggestion, since it increases the alignment requirement for storing metadata.

My point was that your suggestion does not work without increasing memory usage. Your suggestion requires additional changes to fix the alignment, which is what I implemented.

Reply:
It doesn't increase memory usage though; it's functionally identical to the original code. The change you applied is not the same as what I suggested, where you are casting to `**Batch`; my suggestion is effectively your change, but casting to `*align(1) *Batch`.